pkg/controller/managedclusteraddon/managed_cluster_addon_registration.go

```diff
@@ -0,0 +1,279 @@
+	objs := make([]runtime.Object, 0)
+	objs = append(objs, role)
+	objs = append(objs, rolebinding)
+	var retErr error
+	retErr = nil
+	for _, o := range objs {
+		if err := client.Get(
+			context.TODO(),
+			types.NamespacedName{
+				Name:      name,
+				Namespace: klusterletaddonconfig.Namespace,
+			}, o); err != nil && errors.IsNotFound(err) {
+			continue
+		} else if err != nil {
+			retErr = err
+			continue
+		}
+
+		// verify ownerRef
+		if objMetaAccessor, ok := o.(metav1.ObjectMetaAccessor); !ok {
+			log.V(2).Info("Failed to get ObjectMeta")
+			continue
+		} else {
+			ownerRef := metav1.GetControllerOf(objMetaAccessor.GetObjectMeta())
+			if ownerRef == nil {
+				log.V(2).Info("No controller reference of the role, skipping")
+				continue
+			}
+			ownerGV, err := schema.ParseGroupVersion(ownerRef.APIVersion)
+			if err != nil {
+				log.V(2).Info("Failed to get group from object ownerRef")
+				continue
+			}
+			if ownerRef.Kind != klusterletaddonconfig.Kind ||
+				ownerRef.Name != klusterletaddonconfig.Name ||
+				ownerGV.Group != klusterletaddonconfig.GroupVersionKind().Group {
+				log.V(2).Info("Object is not owned by klusterletaddonconfig. Skipping")
+				continue
+			}
+		}
+
+		if err := client.Delete(context.TODO(), o); err != nil && !errors.IsNotFound(err) {
+			retErr = err
+			continue
+		}
+	}
+
+	return retErr
+}
+
+func deleteOutDatedServiceAccount(
+	addon addons.KlusterletAddon,
+	klusterletaddonconfig *agentv1.KlusterletAddonConfig,
+	client client.Client,
+) error {
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      klusterletaddonconfig.Name + "-" + addon.GetAddonName(),
+			Namespace: klusterletaddonconfig.Namespace,
+		},
+	}
+	if err := client.Delete(context.TODO(), sa); !errors.IsNotFound(err) {
+		return err
+	}
+	return nil
+}
+
+func deleteHubKubeConfigResources(
+	addon addons.KlusterletAddon,
+	managedClusterName string,
+	client client.Client) error {
+	componentName := addon.GetAddonName()
+
+	//Create the values for the yamls
+	config := struct {
+		ManagedClusterName      string
+		ManagedClusterNamespace string
+		ComponentName           string
+		ClusterRoleName         string
+	}{
+		ManagedClusterName:      managedClusterName,
+		ManagedClusterNamespace: managedClusterName,
+		ComponentName:           componentName,
+		ClusterRoleName:         addons.GetAddonClusterRolePrefix() + addon.GetManagedClusterAddOnName(),
+	}
+
+	newApplier, err := applier.NewApplier(
+		bindata.NewBindataReader(),
+		&templateprocessor.Options{},
+		client,
+		nil,
+		nil,
+		nil,
+		&applier.Options{
+			Backoff: &wait.Backoff{
+				Steps:    1,
+				Duration: 10 * time.Millisecond,
+				Factor:   1.0,
+			},
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	err = newApplier.DeleteInPath(
+		"resources/hub/common",
+		nil,
+		false,
+		config,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// UpdateRegisrationConfig adds registration annotations to ManagedClusterAddOn
```

Comment:
add a todo to reflect new apis.

Reply:
If possible I'd like to update this pr to use the new apis.